enhancements/ingress/haproxy-router-compression.md

```diff
@@ -0,0 +1,322 @@
+backend server does not support HTTP compression, and the header "Accept-Encoding" exists in the request, HAProxy will
+compress the matching response.  Also, if the HAProxy doesn't have enough resources to perform compression, it may skip
+the compression in order to stay functional.  Process-level directives exist to tune this behavior: `maxcomprate`,
+`maxcompcpuusage`, and `tune.comp.maxlevel` as described in the
+[HAProxy compression documentation](https://www.haproxy.com/documentation/hapee/latest/load-balancing/compression/),
+but these are not in scope of this proposal.
+
+### Risks and Mitigations
+
+Overuse of compression can cause the HAProxy to devote more resources to compression than to other operations,
+compromising its operation.  We do not have any evidence that this will occur, and give detailed notes on use of
+necessary MIME types to mitigate pointless compression and overuse of compression.  Should this become a problem
+in the future, we can add default values for the process-level directives mentioned above (`maxcomprate`,
+`maxcompcpuusage`, and `tune.comp.maxlevel`)  The team consensus is that adding more tunable knobs at this stage is
+premature and could cause more issues than it mitigates.
+
+We considered disallowing wildcard MIME types (e.g. `application/*`), so that explicit choices must be made on which
+MIME types are subject to compression, but as mentioned, it is more likely that any MIME type omission would generate
+a new RFE than for there to be a support issue caused by excessive pointless compression.
+
+Security and UX are not a part of this proposal, but API team members will be invited to comment.
+
+## Design Details
+Design details are covered in the Proposal section, but some additional detailed notes from the
+[HAProxy compression documentation](https://www.haproxy.com/documentation/hapee/latest/load-balancing/compression/)
+are listed below.
+
+### Compression types
+MIME types to apply compression to,  are read from the Content-Type response header field. If a MIME type is not set,
+HAProxy tries to compress all responses that are not compressed by the server.  Therefore, at least one MIME type is
+required in the `HTTPCompressionPolicy`.  We consider the fact that cluster admins may want to create and use their
+own MIME type, but the allowed set of characters is defined in a kubebuilder validation
+pattern  of `"[a-zA-Z0-9\\+-;= ]+/([a-zA-Z0-9\\\+-;= ]+|\\*)"`, to allow a token of alphanumeric plus "_-;= " followed
+by "/" followed by another alphanumeric token of the same pattern, or "*", the wildcard character.
+
+### Open Questions
+Is the MIME type pattern still too strict?  I reviewed the `mime` golang package at https://golang.org/src/mime/type.go,
+`builtinTypesLower` map, but if we intend to allow any custom MIME type, perhaps a wider net should be cast, as in
+https://golang.org/src/mime/grammar.go.
```

Comment:
https://datatracker.ietf.org/doc/html/rfc1341#page-7 provides BNF definitions for allowed MIME types.  If we want to make it a little looser (e.g., allowing non-standard types without the `x-` prefix), I don't object.

Reply:
https://datatracker.ietf.org/doc/html/rfc1521#page-68 is more current.  Should we go that far?